stage #0/module #4 Algorithms, Data Structures/algorithms_and_data_structures.md

```diff
@@ -0,0 +1,487 @@
+In order to define an array reference, 
+1) we should just write a desired type with empty square brackets:
+```
+int[] arrayOfIntegers; // declares an array of integers
+```
+2) we can also place the square-brackets after the identifier:
+```
+int arrayOfIntegers[]; // declares an array of integers
+```
+
+The above two definitions of an arrays have exactly the same meaning.
+But defining an array doesn't mean that we created an array, when array creation means allocating space in memory for storing elements. In order to create an array, we must write an initialization expression.
+
+#### Creation and Initialization
+There are several ways to create an array in Java:
+1) the most generic way is to use an operator "new" followed by a desired type with a desired array sized surrounded by square brackets
+
+```
+int[] arr;        // declares an array of integers
+arr = new int[6]; // creates an array of integers
+```
+
+![Array Initialization](img/array_initialization.png)
+
+It is worth noting that arrays of elements, which are of primitive type, are initialized with default values (in our case each element of an array of integers is initialized with zeros).
+
+2) the special creation way is to write a set of values, which are separated from each other by commas and which are surrounded altogether by curly braces
+
+```
+int[] arr;                     // declares an array of integers
+arr = {12, 14, 1, 3, 16, 27};  // creates and initializes an array of integers
+```
+
+![Special Array Initialization](img/array_initialization_curly_braces.png)
+
+In this case while array creation, its elements will be initialized with desired values.
+
+3) the special creation way has an optional extension: before the aforementioned set of elements in curly braces we can write an operator "new" followed by a desired type and empty square-brackets
+
+```
+int[] arr;                              // declares an array of integers
+arr = new int[] {12, 14, 1, 3, 16, 27}; // creates and initializes an array of integers
+```
+
+![Special Array Initialization](img/array_initialization_curly_braces.png)
+
+An array can be defined, created and initialized in one statement, instead of several: definition and initialization.
+```
+int[] arr1 = new int[6];                       // declares, creates an array of integers (also initializes implicitly with zeros, only for primitive types)
+int[] arr2 = {12, 14, 1, 3, 16, 27};           // declares, creates and initializes an array of integers
+int[] arr3 = new int[] {12, 14, 1, 3, 16, 27}; // declares, creates and initializes an array of integers
+```
+
+#### Length 
+All arrays have an intrinsic member (whether they are arrays of objects or arrays of primitives) that you can query to tell us how many elements there are in the array. This member is "length". Length can not be changed. In order to refer the last element of an array, we should use an index "length - 1".
```

Comment:
(for future) maybe adding why the 1st index number starts with 0 is useful (or as an advanced topic or a link to that)